[JENKINS-41516] Add script console listener

meiswjn opened the pull request:
<!-- Comment: 
A great PR typically begins with the line below.
Replace XXXXX with the numeric part of the issue's id you created on JIRA.
Please note that if you want your changes backported into LTS, you will need to create a JIRA ticket for it. Read https://www.jenkins.io/download/lts/#backporting-process for more.
-->
See [JENKINS-41516](https://issues.jenkins.io/browse/JENKINS-41516).

<!-- Comment: 
If the issue is not fully described in the ticket, add more information here (justification, pull request links, etc.).

 * We do not require JIRA issues for minor improvements.
 * Bugfixes should have a JIRA issue (backporting process).
 * Major new features should have a JIRA issue reference.
-->

### Proposed changelog entries

- Add listener for privileged groovy script execution (including script console and CLI).

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.
The changelog entry should be in the imperative mood; e.g., write "do this"/"return that" rather than "does this"/"returns that".
For examples, see: https://www.jenkins.io/changelog/
-->

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate. 
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

Any

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).


<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/6539"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

meiswjn committed:
Set location for listener

meiswjn committed:
Add ScriptListener

meiswjn committed:
Remove debugging statements

meiswjn committed:
Merge branch 'master' into feature/add-script-console-listener

daniel-beck commented:
This should also cover `GroovyCommand` and `GroovyShCommand` (the latter may be especially awkward to log).

What's your plan for unsandboxed pipelines and other unsandboxed script execution involving `script-security`?

mawinter69 commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -4722,6 +4723,7 @@ public static void _doScript(StaplerRequest req, StaplerResponse rsp, RequestDis
             }
 
             try {
+                ScriptListener.fireScriptEvent(req);
```
why pass the complete request? Wouldn't it be enough to just pass the script itself? 
This code is also called when someone uses the script console of an agent. That information is not available from the request. That probably requires passing a reference to the computer instead of the Channel

meiswjn replied:
Good point. It would be nice to know on what machine the script was executed. I will have a look. Worst case will be a new / different parameter as you suggested.

And yes, I guess the request does not need to be passed.

meiswjn commented:
> > Do you mean unsandboxed pipelines that run potentially dangerous functions that have been put on the allowlist?
> 
> No, users can choose to disable the sandbox for otherwise sandboxed scripts (requiring admin approval unless an admin configures the script and auto-approves), making the resulting features just as powerful/dangerous as the script console.
> 
> I'm not opposed to a change like this, but it should cover all the common scripting features in Jenkins.

That's a good question that I must think about. Gotta take a look at the script-security plugin, then I may be able to come up with a technical solution.
Any hints are welcome, of course.

daniel-beck commented:
FTR `script-security` is its own thing, but it would be beneficial to have a core PR and corresponding plugin PR prepared, because only one of those would only partially address the problem.

meiswjn committed:
Add listener for CLI

meiswjn committed:
Merge branch 'master' into feature/add-script-console-listener

basil commented:
Our general policy is to review proposed API additions along with the proposed consumers. This can help us better understand the intended use case in order to properly review the API and ensure the API can accommodate both present and future use cases. Is there a PR demonstrating the usage of this proposed API addition?

meiswjn commented:
> Our general policy is to review proposed API additions along with the proposed consumers. This can help us better understand the intended use case in order to properly review the API and ensure the API can accommodate both present and future use cases. Is there a PR demonstrating the usage of this proposed API addition?

Hi Basil,
Sure. Here is my planned usecase: https://github.com/jenkinsci/audit-trail-plugin/pull/72

meiswjn commented:
Hey @basil,
Please let me know if this PR needs further justification.
Thanks!

res0nance reviewed (commented):
The idea seems sound. The current api is fine but for the purposes of audit-trail, would it not be beneficial to log the user who executed the script? I'm merely asking to prevent us from having an API that needs extension so quickly

meiswjn commented:
Hi,
I will be very busy for the next two weeks and probably not able to work on this PR during that time. Please feel free to continue reviewing but I may be slow to respond. Especially, I would like to test @res0nance suggestion. Thanks for your understanding.

meiswjn committed:
Merge branch 'jenkinsci:master' into feature/add-script-console-listener

meiswjn committed:
Add new listener for other sources

meiswjn committed:
Simplify listener

meiswjn committed:
Merge branch 'master' into feature/add-script-console-listener

meiswjn committed:
Fix whitespace and comments

meiswjn committed:
Merge branch 'feature/add-script-console-listener' of https://github.com/meiswjn/jenkins into feature/add-script-console-listener

res0nance reviewed (approved):
LGTM!

meiswjn commented:
Is this PR ready for merge? :)

meiswjn commented:
> Please add a changelog entry

Done

daniel-beck commented:
I know I've been absent from this PR for a bit, but would like an opportunity to review it. I do not expect that'll delay it for more than a week. Thanks :)

daniel-beck committed:
Merge branch 'master' into feature/add-script-console-listener

meiswjn commented:
> I know I've been absent from this PR for a bit, but would like an opportunity to review it. I do not expect that'll delay it for more than a week. Thanks :)

Sure. I appreciate your input.

daniel-beck committed:
Merge branch 'master' into feature/add-script-console-listener

daniel-beck commented on core/src/main/java/hudson/cli/GroovyshCommand.java:
```diff
@@ -79,6 +81,7 @@ protected int run() {
         }
 
         Groovysh shell = createShell(stdin, stdout, stderr);
+        ScriptListener.fireScriptEvent(commandLine.toString(), "CLI", User.current());
```
This does not actually record the script submitted via the shell by the user. It records the command line arguments (probably none) once, before the user actually starts using the shell. So all this does is record _that_ someone (and who_ used the `groovysh` command, but nothing about the actual script commands.

daniel-beck replied:
Something like https://github.com/daniel-beck/jenkins/commit/9590e638db4f16877895211fb2db726363f6f94b may do it.

I have not yet figured out how the Groovysh profile files work; those could potentially still be used to bypass logging.

meiswjn replied:
> This does not actually record the script submitted via the shell by the user. It records the command line arguments (probably none) once, before the user actually starts using the shell. So all this does is record _that_ someone (and who) used the `groovysh` command, but nothing about the actual script commands.



> Something like [daniel-beck@9590e63](https://github.com/daniel-beck/jenkins/commit/9590e638db4f16877895211fb2db726363f6f94b) may do it.
> 
> I have not yet figured out how the Groovysh profile files work; those could potentially still be used to bypass logging.

Very good find, thank you! I implemented your solution. 

Regarding the profiles: If I understand Groovy Profiles correctly (and my tests weren't completely wrong), the Groovy profile on the client-side would have no effect.

Wadeck commented on core/src/main/java/hudson/cli/GroovyCommand.java:
```diff
@@ -70,7 +72,9 @@ protected int run() throws Exception {
         GroovyShell groovy = new GroovyShell(Jenkins.get().getPluginManager().uberClassLoader, binding);
-        groovy.run(loadScript(), "RemoteClass", remaining.toArray(new String[0]));
+        String script = loadScript();
+        ScriptListener.fireScriptEvent(script, "CLI", User.current());
```
```suggestion
        ScriptListener.fireScriptEvent(script, "CLI/GroovyCommand", User.current());
```

💡 As you include the origin, I imagine you would like to use it for auditing purpose, so having more information between the two CLI calls could be interesting. WDYT?
(And another one for the GroovyshCommand of course)

Wadeck reviewed (commented):
Not manually tested, so no approval from my side, just comment.

> No tests included as no implementation of listener. Must be tested in plugin.

BTW nothing prevents you to implement a dummy listener just for test purpose :) Having no tests means that the likelihood a regression will be introduced is increased.
The last comment from Daniel also shows a non-working feature from this PR, which could have been caught potentially by the tests.

meiswjn committed:
Update core/src/main/java/hudson/cli/GroovyCommand.java

Co-authored-by: Wadeck Follonier <Wadeck@users.noreply.github.com>

meiswjn committed:
Add precise origins; Correctly log CLI input

meiswjn committed:
Merge branch 'jenkinsci:master' into feature/add-script-console-listener

meiswjn committed:
Merge branch 'feature/add-script-console-listener' of https://github.com/meiswjn/jenkins into feature/add-script-console-listener

meiswjn committed:
Add LoggingGroovySh as suggested by @daniel-beck

meiswjn committed:
Add ScriptListener Tests

meiswjn commented:
> Not manually tested, so no approval from my side, just comment.
> 
> > No tests included as no implementation of listener. Must be tested in plugin.
> 
> BTW nothing prevents you to implement a dummy listener just for test purpose :) Having no tests means that the likelihood a regression will be introduced is increased. The last comment from Daniel also shows a non-working feature from this PR, which could have been caught potentially by the tests.

Hi @Wadeck,
Thanks for your feedback. I had tests implemented in the downstream, but for the CLI the tests only covered GroovyCommand, not GroovyShCommand. I fixed it and implemented the tests in this PR as well.

meiswjn commented:
Hey @daniel-beck,
Did you find any time yet to review the changes? Thank you!

meiswjn committed:
Merge branch 'master' into feature/add-script-console-listener

basil committed:
Merge branch 'master' into feature/add-script-console-listener

daniel-beck commented:
@meiswjn I'll do another iteration in the next few days.

dwnusbaum commented:
> I'm not opposed to a change like this, but it should cover all the common scripting features in Jenkins.

For what it's worth, I think that the usage pattern for `ScriptApproval` is different enough that it might not make sense to include it in this API. `ScriptApproval` is intended for persistent scripts rather than ephemeral ones, which already makes it more auditable via things like `jobConfigHistory`, it is exposed to unprivileged users in regular features via classes like `SecureGroovyScript`, and it makes a distinction between script approval and script execution. A listener that fires whenever an admin runs an ad-hoc Groovy script in the script console or via the CLI should fire relatively rarely and only have "interesting" events, whereas firing an event for every execution of a`SecureGroovyScript` seems unhelpful. Maybe it would make sense to only fire an event when `ScriptApproval.approveScript` is called, but either way I think that `ScriptApproval` is quite different from the script console, and something like https://github.com/jenkinsci/script-security-plugin/pull/300 would make more sense for auditing its usage.

daniel-beck commented:
> whereas firing an event for every execution of a `SecureGroovyScript` seems unhelpful

FWIW https://github.com/jenkinsci/script-security-plugin/pull/416#pullrequestreview-1056562733

daniel-beck committed:
Add basic default listener that just logs

daniel-beck commented:
Some notes while I'm still playing with this (not immediately actionable I think):

* Would probably be better if "origin" was an object; could pass the `Run` directly rather than a string representation.
* Ideally this would also record the script binding. Depending on the script, a lot can be done with the variables passed via binding, and not having a chance to log those seems like a gap.

meiswjn commented:
> > I'm not opposed to a change like this, but it should cover all the common scripting features in Jenkins.
> 
> For what it's worth, I think that the usage pattern for `ScriptApproval` is different enough that it might not make sense to include it in this API. `ScriptApproval` is intended for persistent scripts rather than ephemeral ones, which already makes it more auditable via things like `jobConfigHistory`, it is exposed to unprivileged users in regular features via classes like `SecureGroovyScript`, and it makes a distinction between script approval and script execution. A listener that fires whenever an admin runs an ad-hoc Groovy script in the script console or via the CLI should fire relatively rarely and only have "interesting" events, whereas firing an event for every execution of a`SecureGroovyScript` seems unhelpful. Maybe it would make sense to only fire an event when `ScriptApproval.approveScript` is called, but either way I think that `ScriptApproval` is quite different from the script console, and something like [jenkinsci/script-security-plugin#300](https://github.com/jenkinsci/script-security-plugin/pull/300) would make more sense for auditing its usage.

Thanks for your feedback! My goal was to track every usage of privileged script execution - even by admins and approved scripts. For example, I do want to know when anyone uses `hudson.util.Secret.decrypt` and not just when it is approved for the first time. Having audit logs for potentially uninteresting events is not a problem from my perspective.

daniel-beck commented:
Some more notes while I'm digging into this further:

* Groovy hook scripts should be logged.
* We probably want to log the result and/or output of the script evaluation, where applicable.

daniel-beck commented:
> I have not yet figured out how the Groovysh profile files work; those could potentially still be used to bypass logging.

https://github.com/apache/groovy/blob/GROOVY_2_4_21/subprojects/groovy-groovysh/src/main/groovy/org/codehaus/groovy/tools/shell/Groovysh.groovy#L397 calls https://github.com/apache/groovy/blob/GROOVY_2_4_21/subprojects/groovy-groovysh/src/main/groovy/org/codehaus/groovy/tools/shell/commands/LoadCommand.groovy#L88 calls https://github.com/apache/groovy/blob/GROOVY_2_4_21/subprojects/groovy-groovysh/src/main/groovy/org/codehaus/groovy/tools/shell/Shell.groovy#L122 calls https://github.com/apache/groovy/blob/GROOVY_2_4_21/subprojects/groovy-groovysh/src/main/groovy/org/codehaus/groovy/tools/shell/Groovysh.groovy#L158 so `LoggingGroovySh` should cover this (untested).

----

FTR https://github.com/daniel-beck/jenkins/tree/meiswjn-feature-add-script-console-listener is my current work in progress. The API got changed incompatibly (and will change again because what I did was wrong), so there's currently no working `script-security` integration yet.

meiswjn commented:
Hi Daniel, thanks for all the feedback! I will wait with responding until you are done :)

Just one thing in between: I am not sure if we should log the output. The usecase that I wanted this feature for was logging when admins decrypt secrets via the console. We would need to ensure that sensitive values do not appear in the audit log. 

Logging the output seems like it adds plenty of complexity. Should we view that as two isolated features, perhaps?

daniel-beck commented:
> The usecase that I wanted this feature for was logging when admins decrypt secrets via the console. We would need to ensure that sensitive values do not appear in the audit log.

That's what

> and will change again because it becomes apparent that there need to be more methods if we start logging output as well

hints at; output would be in something like `onScriptOutput` since the existing parameters don't make a lot of sense except to create correlation of some kind. Leave that unimplemented, and you won't see output.

Or are you saying nobody should be able to record any output?

meiswjn commented:
I would say that recording the script's output is a feature that could be considered, but I would say it is not as important as recording the script itself. I would suggest it is out of scope for this PR. Low reward, high effort from my perspective.

daniel-beck commented:
> Low reward, high effort from my perspective.

Mostly done :) (unless we count pipelines, those would probably be pretty painful; but their output is recorded elsewhere already anyway)

FWIW I realize I'm a bit more involved here than in most other PRs as just a reviewer. As you can see in the older Jira issue I updated this PR to, I've been involved in occasional discussions around this general topic for more than five years. If you prefer, I'll step back and let you finish this.

meiswjn commented:
If you want to implement it, sure! Feel free :)

meiswjn commented:
Hey @daniel-beck, I know you are pretty busy but I would appreciate if you could quickly lay out your plans for this PR :) Do you want to merge this one or yours?

meiswjn committed:
Merge branch 'master' into feature/add-script-console-listener

daniel-beck commented:
@meiswjn Sorry for the delay in responding, unfortunately I am getting too many notifications to handle these days.

This PR, as is, is not mergeable IMO. Whether you pull in changes from #7056, or we just "migrate" this project over there, does not really matter to me.

FWIW this is still near the top of my TODO list, and I'm hoping to have something mergeable this month.

meiswjn committed:
Merge branch 'master' into feature/add-script-console-listener

meiswjn commented:
Superseded by #7056
